docs/set.rst

```diff
@@ -0,0 +1,487 @@
+
+::
+
+    Set.intersection Set.empty (Set.fromList [0, 2, 4, 6])
+    > fromList []
+
+    Set.intersection (Set.fromList [1, 3, 5, 7]) (Set.fromList [0, 2, 4, 6])
+    > fromList []
+
+    Set.intersection (Set.singleton 0) (Set.fromList [0, 2, 4, 6])
+    > fromList [0]
+
+Difference
+""""""""""
+
+::
+
+    Set.difference :: Ord a => Set a -> Set a -> Set a
+    Set.difference l r = ...
+
+:set:`difference` returns a set containing the elements that are in the first
+set ``l`` but not the second set ``r`` (`set difference/relative compliment
+<https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement>`_).
+
+::
+
+    Set.difference (Set.fromList [0, 2, 4, 6]) Set.empty
+    > fromList [0,2,4,6]
+
+    Set.difference (Set.fromList [0, 2, 4, 6]) (Set.fromList [1, 3, 5, 7])
+    > fromList [0,2,4,6]
+
+    Set.difference (Set.fromList [0, 2, 4, 6]) (Set.singleton 0)
+    > fromList [2,4,6]
+
+Subset
+""""""
+
+::
+
+    Set.isSubsetOf :: Ord a => Set a -> Set a -> Bool
+    Set.isSubsetOf l r = ...
+
+:set:`isSubsetOf` returns ``True`` if all elements in the first set ``l`` are
+also in the second set ``r`` (`subset <https://en.wikipedia.org/wiki/Subset>`_).
+
+.. NOTE::
+   We use `infix notation
+   <https://wiki.haskell.org/Infix_operator#Using_infix_functions_with_prefix_notation>`_
+   so that it reads nicer. These are back-ticks (`), not single quotes (').
+
+::
+
+    Set.empty `Set.isSubsetOf` Set.empty
+    > True
+
+    Set.empty `Set.isSubsetOf` (Set.fromList [0, 2, 4, 6])
+    > True
+
+    (Set.singleton 0) `Set.isSubsetOf` (Set.fromList [0, 2, 4, 6])
+    > True
+
+    (Set.singleton 1) `Set.isSubsetOf` (Set.fromList [0, 2, 4, 6])
+    > False
+
+
+Typeclass Instances
+-------------------
+
+``Set`` is an instance of a number of common typeclasses, for the full list see
+the `docs
+<https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-Set.html#t:Set>`_.
+
+.. NOTE::
+   Some constraints have been left out for brevity, and the types given below
+   are speciliazed to ``Set``; the true types are more general.
+
+- `Show
+  <https://hackage.haskell.org/package/base-4.10.1.0/docs/Prelude.html#t:Show>`_ -
+  conversion to string: ``show :: Show a => Set a -> String``
+- `Eq
+  <https://hackage.haskell.org/package/base-4.10.1.0/docs/Prelude.html#t:Eq>`_ -
+  equality check: ``(==) :: Eq a => Set a -> Set a -> Bool``
+- `Ord
+  <https://hackage.haskell.org/package/base-4.10.1.0/docs/Prelude.html#t:Ord>`_ -
+  comparison: ``(<) :: Ord a => Set a -> Set a -> Bool``
+- `Foldable <https://wiki.haskell.org/Typeclassopedia#Foldable>`_ - collapse
+  into summary value: ``foldr :: (a -> b -> b) -> b -> Set a -> b``
+- `Semigroup <https://wiki.haskell.org/Typeclassopedia#Semigroup>`_ - combine
+  two things together (`union`_): ``(<>) :: Set a -> Set a -> Seq a``
+- `Monoid <https://wiki.haskell.org/Typeclassopedia#Monoid>`_  - a semigroup
+  with an identity element: ``mempty :: Seq a``
```

Comment:
I don't think explaining what the classes are is too helpful. Explaining how they apply to `Set` and how they're used with sets might be useful.

Reply:
Removed the typeclasses section. I think it'll require some more thought to figure out how to present them best which I'd like to do as a follow-up.